README.md
Callimachus Project Installation Instructions
=============================================

http://callimachusproject.org/

----------------
| Introduction |
----------------

Thank you for your interest in the Callimachus Project.  This is the README
for Callimachus (Open Source), a Linked Data management system.

Callimachus aims to make Semantic Web applications (much) easier to create.
Because this is a community effort, you can help by submitting your
applications, upgrades and tools to the Callimachus Project site at
http://callimachusproject.org.

Documentation for Callimachus may be found at:
  http://callimachusproject.org/documentation.xhtml?view

Please consider joining the mailing lists and participating in the
online community so we can serve you better.  See
http://callimachusproject.org for more details.

This software is licensed as defined in the accompanying files
LICENSE.txt and LEGAL.txt.  These files may be found in the
installation directory.

-----------------
| prerequisites |
-----------------

You must have Java JDK 7 installed and available with ECMAScript 5
support. Oracle introduced EMCAScript 5 support in Java™ SE 7 Update 1.

Callimachus is expected to work on the following server platforms:
* Linux i386
* Linux 64bit
* Mac on Intel
* Windows XP
* Windows Vista
* Windows 7

-----------------
| Configuration |
-----------------

An SMTP server is required to send an invite email and password and reset user's password.
The SMTP server can be configured in the file etc/mail.properties.
A typical file might contain the following block:

    mail.transport.protocol = smtp
    mail.from = system@example.com
    mail.host = example.com
    mail.smtp.port = 25
    mail.smtp.auth = true
    mail.smtp.starttls.enable = true
    mail.smtp.starttls.required = true
    mail.user = system
    mail.password = secret

For more options see:
> http://java.sun.com/javaee/5/docs/api/javax/mail/package-summary.html
> http://java.sun.com/products/javamail/javadocs/com/sun/mail/smtp/package-summary.html


----------------
| Installation |
----------------

There are two ways to acquire and run Callimachus:  Download the ZIP archive 
or checkout the code repository.

If you download a ZIP archive of a release from http://callimachusproject.org/, 
extract it into a new directory. Copy etc/callimachus-defaults.conf to
etc/callimachus.conf; change the ORIGIN and PORT to the be the same as the
hostname and port (if not port 80). Be sure to remove the preceding comment
character ('#') in the .conf file.

To register a Callimachus daemon with the Linux system (to stop Callimachus
on machine shut down) run the bin/callimachus-install.sh script. The
Callimachus daemon is only available for Linux. As part of the installation, the
script will ask to generate a server certificate. This certificate is used for
all https origins and for remote JMX administration.

Execute a callimachus-setup script located in the bin/ directory to initialize
the repository. Open your Web browser to http://localhost:8080/ to access the
service. Execute a callimachus-stop script located in the bin/ directory to
stop the server. To start the server again use the provided start script.

On Linux server run (as root):

    # bin/callimachus-install.sh
    # bin/callimachus-setup.sh

On Linux or Mac desktop run:

    $ bin/callimachus-setup.sh

On Windows desktop run:

    # bin/callimachus-setup.bat

To monitor the activity of the server watch the log/callimachus.log.0 file for
log messages. Windows users may find the callimachus-log.bat script useful to
monitor the activity without locking the log file.


Alternatively, if you want to check out the source code, you can do so here on
GitHub via the command git clone https://github.com/3-Round-Stones/callimachus.git.
Once you have the source code checked out, you will need to provide a username,
email and password. Create a build.properties file in the same directory as
build.xml. Put the variables "callimachus.username", "callimachus.email", and
"callimachus.password" followed by '=' and their value on their own line. Copy
src/callimachus-defaults.conf to etc/callimachus.conf and edit file as above.
You can then execute 'ant run' from a command line in this top directory to run
the server. If some files fail to download you may have to run "ant clean dist"
and run "ant run" again.

To test the build run "ant test".

Additional documentation regarding usage and application development may
be found on the project's wiki at:
>  http://callimachusproject.org/getting-started-with-callimachus

Details of browser support may be found at:
>  http://callimachusproject.org/docs/1.2/articles/browser-support.docbook?view

Share your experience or ask questions on the mailing list:
callimachus-discuss@googlegroups.com.  You can sign up at 
>  http://groups.google.com/group/callimachus-discuss.



RELEASE_NOTES.md
# Callimachus Project Release Notes

> http://callimachusproject.org/

10 February 2016

## REQUIREMENTS FOR CALLIMACHUS 1.5.0 (Open Source)

Callimachus works best on browsers that are closely tracking the development
of HTML5.  At the time of this writing, Callimachus works best on Chrome and
Firefox.  It is recommended that Callimachus users track updates to their
browsers for the best usage experience.

Chrome 48 has been thoroughly tested.
Firefox 44 has no known issues.
Safari 6.1 and 7 has known issues.
Internet Explorer 11 has known issues.

Details of browser support may be found at:
  http://callimachusproject.org/docs/1.2/articles/browser-support.docbook?view

The Callimachus server requires Java JDK 1.7 on the server to run.

## NEW IN VERSION 1.5.0

 * New Remote RDF Datasource resources provide Direct and Indirect Graph Store Protocol and navigable resources on top of (remote) SPARQL endpoints http://www.w3.org/TR/sparql11-http-rdf-update/
 * Support for text/turtle and application/ld+json result formats from SPARQL endpoints http://www.w3.org/TR/sparql11-protocol/
 * Support for application/sparql-update file hosting
 * SERVICE blocks now use Callimachus credentials for authentication http://callimachusproject.org/getting-started-with-callimachus#Home_Folder_Edit_Tab
 * Faster CAR folder import
 * Member of staff and power groups can now invite users to groups they are editors of
 * Folders can now define their own 404 handlers
 * New Persistent URL resources https://en.wikipedia.org/wiki/Persistent_uniform_resource_locator
   * Redirect resources simplify the PURL creation be providing only essential fields
   * PURL interface has been simplified to provide GET-only capabilities
   * Proxy resources proxy content and operations by a path prefix from the Web
   * Rewrite Rule allows complex redirection/handling of requests for resources of any path depth

## KNOWN ISSUES IN 1.5.0

 * When resource labels are updated, referencing resource may not display new
   label until they are modified as well or the callimachus-reset script is run.
 * Template variables are only bound if there is at least one triple binding after them.
 * On Windows systems with certain installations of Git you may see an error on startup e.g. 'couldn't find "#"'. If this happens, make sure that the System32 folder is before Git's tools folder on the path so that the correct version of "find" is found.

## HOWTO UPGRADE

Stop the server using the callimachus-stop script in the bin directory.
Remove the lib, bin, tmp, src directories.

Download and unzip the callimachus zip archive file to the install directory of
the Callimachus instance to be upgraded. Run the bin/callimachus-setup script to
upgrade the repositories directory. Use the "-K" flag to disabled automatic
backup.

Templates using the 1.3 syntax must be updated to use the new calli. functions
available in 1.4.0 and 1.5.0.
 * Links to <aClass?create> must be changed to `<aFolder?create=aClass>` syntax
 * Create forms must be changed to have `enctype="text/turtle"`
 * Create forms must now have
     `onsubmit="calli.submitTurtle(event,calli.slugify($('#label').val()))"`
 * Edit template body tags must be
     `<body resource="?this" onload="comparison=calli.copyResourceData('#form')">`
     Where "form" is the @id of the form tag.
 * Edit forms must have `onsubmit="calli.submitUpdate(comparison,event)"`
 * Both create and edit forms' input and textarea tags must include
     `onchange="calli.updateProperty(event, 'rdfs:label')"`
     Where rdfs:label is the datatype property of the field
 * `<select/>`, `type="checkbox"`, and `type="radio"` fields that are populated from
     the RDF store must be replaced with an RDF Named Query and included using
     `<xi:include href="select-query.rq?select" />` or another pragrma
     ?select, ?checkbox, ?checkbox-include, ?radio, ?radio-inline.
     The following (and selectEachResourceIn) copies RDFa data to a field:
     `$('[rel="dcterms:type"].hidden').each(calli.checkEachResourceIn('#type')).remove();`
 * View templates must replace `<pre class="wiki" property="rdfs:comment" />` with
     `<p property="rdfs:comment" />` if no wiki syntax is used or
```
    <pre property="rdfs:comment" />
    <script type="text/javascript">$(function($){
        $('pre[property]').replaceWith(calli.parseCreole);
    });</script>
```
 * Replace `<iframe class="flex"/>` with `calli.fillElement('iframe')`
 * Replace `<aside class="optional"/>` with `$('aside').filter(calli.isEmptyResource).remove();`
 * Remove class attributes from `<time/>` and use [calli.parseDateTime](http://callimachusproject.org/callimachus-reference#parseDateTime) to format
 * calliSubmit, calliRedirect and other calli* events must be replaced with [a function call](http://callimachusproject.org/callimachus-reference#JavaScript_Reference)
 * When [creating RDF resources using the Callimachus REST API](http://callimachusproject.org/callimachus-reference#RDF_Create), use text/turtle instead of sparql-update
 * Links with class "view" and "diverted" have no special behaviour
 



webapp/scripts/purl.js
// purl.js
/*
   Copyright (c) 2014 3 Round Stones Inc., Some Rights Reserved
   Licensed under the Apache License, Version 2.0, http://www.apache.org/licenses/LICENSE-2.0
*/
jQuery(function($){
    var mode = getMode();   // get the subclass of PURL being created/edited
    
    var actionTypeOptions = [
        { value: 'calli:copy', text: 'Copy (200)' },
        { value: 'calli:canonical', text: 'Canonical (301)' },
        { value: 'calli:alternate', text: 'Alternate (302)' },
        { value: 'calli:describedby', text: 'Described by (303)' },
        { value: 'calli:resides', text: 'Resides (307)' },
        { value: 'calli:moved', text: 'Moved (308)' },
        { value: 'calli:missing', text: 'Missing (404)' },
        { value: 'calli:gone', text: 'Gone (410)' },
        { value: 'calli:illegal', text: 'Illegal (451)' },
        { value: 'calli:post', text: 'POST' },
        { value: 'calli:put', text: 'PUT' },
        { value: 'calli:patch', text: 'PATCH' },
        { value: 'calli:delete', text: 'DELETE' }
    ];
    
    var redirectActions = {
        '301': actionTypeOptions[1],
        '302': actionTypeOptions[2]
    };
    
    var purlActions = {
        '200': actionTypeOptions[0],
        '301': actionTypeOptions[1],
        '302': actionTypeOptions[2],
        '303': actionTypeOptions[3],
        '307': actionTypeOptions[4],
        '308': actionTypeOptions[5],
        '404': actionTypeOptions[6],
        '410': actionTypeOptions[7],
        '451': actionTypeOptions[8]
    };

    // check for limited action types in the query string
    var params = parseQueryString();
    var allowedActions = [];
    if ((mode === 'Redirect' || mode === 'Purl') && (params.action !== undefined || params.actions !== undefined)) {
        var val = params.actions || params.action;
        if (typeof val === 'string') {
            val = val.split(',');
        }
        var list = (mode === 'Redirect') ? redirectActions : purlActions;
        for (var i in val) {
            if (val[i] in list) {
                allowedActions.push(list[val[i]]);
            }
        }
        removeDefinitions(allowedActions);
    }

    if (mode === 'Proxy') {
        // set up collapsible panels in Proxy templates
        $(".collapse").on("hide.bs.collapse", function(){
            $(this).parent().find('.panel-heading a').removeClass('glyphicon-minus').addClass('glyphicon-plus');
            $(this).find('input[type="hidden"]').removeAttr('property');
        });
        $(".collapse").on("show.bs.collapse", function(){
            $(this).parent().find('.panel-heading a').removeClass('glyphicon-plus').addClass('glyphicon-minus');
        });
    }

    $('#form').each(function(){
        var form = this;
        var comparison = calli.copyResourceData(form);
        $(form).submit(function(event){
            event.preventDefault();
            calli.resolve($(form).attr("enctype")).then(function(enctype){
                var errors = [];
                validate(mode, enctype, errors);
                if (errors.length !== 0) {
                    return calli.reject($('<span/>').append('<b>Validation errors:</b><br>'+errors.join('<br>'))[0]);
                }

                var action = calli.getFormAction(form);
                if (enctype == "text/turtle") {
                    return calli.resolve(form).then(function(form){
                        var local = $('#local').val();
                        var ns = window.location.pathname.replace(/\/?$/, '/');
                        var resource = ns + encodeURI(local).replace(/%25(\w\w)/g, '%$1').replace(/%20/g, '+');
                        form.setAttribute("resource", resource);
                        return calli.copyResourceData(form);
                    }).then(function(data){
                        data.results.bindings.push({
                            s: {type:'uri', value: data.head.link[0]},
                            p: {type:'uri', value: 'http://purl.org/dc/terms/created'},
                            o: {
                                type:'literal',
                                value: new Date().toISOString(),
                                datatype: "http://www.w3.org/2001/XMLSchema#dateTime"
                            }
                        });
                        return data;
                    }).then(function(data){
                        return calli.postTurtle(action, data);
                    });
                } else {
                    return calli.resolve(form).then(calli.copyResourceData).then(function(insertData){
                        insertData.results.bindings.push({
                            s: {type:'uri', value: insertData.head.link[0]},
                            p: {type:'uri', value: 'http://purl.org/dc/terms/modified'},
                            o: {
                                type:'literal',
                                value: new Date().toISOString(),
                                datatype: "http://www.w3.org/2001/XMLSchema#dateTime"
                            }
                        });
                        return insertData;
                    }).then(function(insertData){
                        return calli.postUpdate(action, comparison, insertData);
                    });
    
                }
            }).then(function(redirect){
                if (window.parent != window && parent.postMessage) {
                    parent.postMessage('POST resource\n\n' + redirect, '*');
                }
                window.location.replace(redirect + '?view');
            }, calli.loading(event.target, calli.error));
        });
    });
    
    // load PURL rules from the resource into the edit space
    $('#loadedRules input').each(function() {
        var purl = decodePurl($(this).attr('property'), $(this).val().trim());
        var rule = null;
        if ($('#blankRule').length){
            // multiple rule table row
            rule = $('#blankRule>div').clone();
            $('#rules').append(rule);
        } else if ($('#rules div.rule[data-method]').length) {
            if (purl.action === 'calli:alternate') {
                // ignore the calli:alternate rule that is a special view case when there is no GET
                $(this).remove();
                return;
            }
            rule = $('#rules').find('div.rule[data-method="'+purl.methods+'"]');
            if (rule.length === 0) {
                $(this).remove();
                return;
            }
            rule.find('.panel-body').collapse('show');
            rule.find('input[type="hidden"]').remove(); // remove dummy rule
        } else {
            // single rule inputs
            rule = $('#rules');
        }
        rule.find('.requestMethod').val(purl.methods.split(','));
        rule.find('.requestPattern').val(purl.pattern);
        rule.find('.actionType').val(purl.action);
        rule.find('.purlTarget').val(purl.template);
        rule.find('.purlTarget').after($(this));
    });

    // load PURL rules from the resource into the view table
    $('#loadedRules p').each(function() {
        var purl = decodePurl($(this).attr('property'), $(this).attr('content').trim());
        if (mode !== 'Proxy' || purl.action !== 'calli:alternate') {
            // ignore the alternate rule as itis just for viewing when there is no GET rule
            $('#rules').append('<tr><td>'+purl.methods+'</td><td>'+(purl.pattern?purl.pattern:'')+'</td><td>'+$(this).text()+'</td><td><a href="'+purl.template+'">'+purl.template+'</a></td></tr>');
        }
    });

    // load a PURL rule from the resource into the view table for simple PURLs
    $('#loadedRule p').each(function() {
        var purl = decodePurl($(this).attr('property'), $(this).attr('content').trim());
        $('#rule').append(
            '<tr><th>Incoming request</th><td>'+purl.methods+'</td></tr>'+
            '<tr><th>Action target</th><td><a href="'+purl.template+'">'+purl.template+'</a></td></tr>'+
            '<tr><th>Status</th><td>'+$(this).text()+'</td></tr>'
        );
    });

    // set up first row and any selects
    if (mode === 'Redirect' || mode === 'Purl') {
        setupRuleSelects($('#rules'), mode, allowedActions);
    } else if (mode === 'Proxy') {
        setupRuleSelects($('#rules'), mode);
    } else {
        if ($('#loadedRules').length === 0) {
            // Create the initial row for a PURL rule in a create template
            addRule('#rules', mode);
            $('#rules a.glyphicon-minus').hide();
        } else {
            setupRuleSelects($('#rules'), mode);
        }
    }

    if ($('#cache').length) {
        var selectionValue = $('#cache>option:selected').val();
        if (selectionValue && $('#cache>option[value="'+selectionValue+'"]').length === 2) {
            $('#cache>option:selected').remove();
            $('#cache').val(selectionValue);
        }
        $('#cache').selectize({
            allowEmptyOption:true,
            create:true
        });
    }
});

function getMode() {
    var mode = $('#form').attr('typeof'); // only present in a create template
    if (mode) {
        mode = mode.split(':')[1];
    } else {
        mode = $('div[property="rdf:type"]:contains("http://callimachusproject.org/rdf/2009/framework")').text();
        mode = mode.split('#')[1];
    }
    return mode;
}

function decodePurl(actionType, target) {
    var purl = {};
    var res = target.match(/^((?:(?:GET|PUT|POST|PATCH|DELETE),?)*)\s*(?:(\S+)\s+)?\s*(.*)$/);
    purl.methods = (res[1] && res[1].length !== 0)? res[1] : null;
    purl.pattern = res[2];
    purl.template = res[3] || null;
    purl.action = actionType;
    if (purl.methods === null) {
        if (actionType === 'calli:post') { 
            purl.methods = 'POST'; 
        } else if (actionType === 'calli:patch') { 
            purl.methods = 'PATCH';
        } else if (actionType === 'calli:put') { 
            purl.methods = 'PUT';
        } else if (actionType === 'calli:delete') { 
            purl.methods = 'DELETE';
        } else { 
            purl.methods = 'GET';
        }
    }
    return purl;
}

// validate the rules and purl definition
function validate(mode, enctype, errors) {
    /*
    General
        Body allowed for POST, PUT, PATCH
        Combination of regex and request methods must not overlap - not easy to test
    RewriteRule
        No further restriction
    Proxy
        1 rule per request method
        Simplified Cache-Control
    PURL
        Single rule
        Not partial
        Simplified Cache-Control
        No pattern matching
        No headers/body - use RewriteRule
    Redirect
        Single rule
        Not partial
        No cache control
        No pattern matching
        No headers/body - use RewriteRule

    Folder - TBD
        1 rule per action type ???
        No headers/body - use RewriteRule
        
    Validation of target template, body, headers not yet implemented
    */
    var getMethodsFound = 0;
    // check the name on a create template
    if (enctype == "text/turtle") {
        var local = $('#local').val().trim();
        $('#local').val(local);
        if (local.charAt(local.length-1) === '/') {
            if (mode === 'Redirect' || mode === 'Purl') {
                errors.push('This type of PURL cannot end with / as they are for simple redirections');
            }
        }
    }
    
    // set up the hidden inputs to represent the full predicate & object pairs of the PURL
    var ruleCount = 0;
    $('#rules input[type="hidden"]').each(function(idx) {
        var rule = $(this).parents('.rule');
        if (mode !== 'Proxy' || rule.find('.panel-body').hasClass("in")) {
            // for a proxy, ignore collapsed panels
            var purl = validateRule(rule, idx, errors);

            // build full template for the rule if it is valid
            if (purl) {
                ruleCount++;
                $(this).attr('property', purl.action);
                if (purl.methods && purl.methods.indexOf('GET') !== -1) {
                    getMethodsFound++;
                }
                $(this).val(encodePurl(purl, mode));
            }
        }
        
    });
    if (ruleCount === 0 && mode === 'Proxy') {
        errors.push('Please define at least one handler');
    }
    if (ruleCount !== 0 && errors.length === 0 && !getMethodsFound && mode !== 'Redirect' && mode !== 'Purl') {
        // if there are no valid PURLs responding to GET requests create a redirect to a view of this PURL
        // what about those with pattern matching that don't resolve to this PURL?
        $('#form').append("<input type='hidden' property='calli:alternate' content='?view' value='?view'/>");
    }
}

function validateRule(rule, idx, errors) {
    var purl = {};
    var ruleError = false;
    // validate user entries
    if (rule.find('.requestMethod').length) {
        purl.methods = rule.find('.requestMethod').val();
    } else if (rule.attr('data-method')) {
        purl.methods = [rule.attr('data-method')];
    } else {
        // this is a fixed method
        purl.methods = ['GET'];
    }
    if (purl.methods === null) {
        errors.push('Rule '+(idx+1)+': Request method is mandatory.');
        ruleError = true;
    }
    if (rule.find('.requestPattern').length) {
        purl.pattern = rule.find('.requestPattern').val().trim();
    } else {
        // Method is GET only so pattern is also required
        purl.pattern = '.*';
    }
    purl.template = rule.find('.purlTarget').val() || null;
    if (purl.template) {
        purl.template = purl.template.trim();
    }
    if (purl.template === null || purl.template === '') {
        errors.push('Rule '+(idx+1)+': Target template is mandatory.');
        ruleError = true;
    }
    var actionElem = rule.find('.actionType');
    purl.action = actionElem.val() || actionElem.attr('data-value');
    if (purl.action === null || purl.action === '') {
        errors.push('Rule '+(idx+1)+': Action method/status is mandatory.');
        ruleError = true;
    }
    if (purl.pattern === null || purl.pattern === '' && patternRequired(purl)) {
        errors.push('Rule '+(idx+1)+': Request pattern is mandatory.');
        ruleError = true;
    }
    return ruleError ? null : purl;
}

function patternRequired(purl) {
    // pattern is required when a method is given or the request method doesn't match the action
    return (purl.methods.length > 1) || !(
        (purl.methods[0] === 'GET' && (
            purl.action === 'calli:copy' || purl.action === 'calli:canonical' || purl.action === 'calli:alternate' || 
            purl.action === 'calli:describedby' || purl.action === 'calli:resides' || purl.action === 'calli:moved' ||
            purl.action === 'calli:missing' || purl.action === 'calli:gone' || purl.action === 'calli:illegal')) ||
        (purl.methods[0] === 'POST' && purl.action === 'calli:post') ||
        (purl.methods[0] === 'PUT' && purl.action === 'calli:put') ||
        (purl.methods[0] === 'PATCH' && purl.action === 'calli:patch') ||
        (purl.methods[0] === 'DELETE' && purl.action === 'calli:delete')
    );
}

function encodePurl(purl, mode) {
    if (mode !== 'Redirect' && mode !== 'Purl') {
        // simplify some classes of PURL
        if (purl.methods && purl.methods.length === 1) {
            // only one rule so it may not be required
            if ((purl.methods[0] === 'POST' && purl.action === 'calli:post') ||
                (purl.methods[0] === 'PUT' && purl.action === 'calli:put') ||
                (purl.methods[0] === 'PATCH' && purl.action === 'calli:patch') ||
                (purl.methods[0] === 'DELETE' && purl.action === 'calli:delete') ||
                (purl.methods[0] === 'GET' && purl.action !== 'calli:post' && purl.action !== 'calli:put' && purl.action !== 'calli:patch' && purl.action !== 'calli:delete')) {
                // this is the default method for this action type so not required in the PURL
                purl.methods = null;
            }
        }
    } else {
        // Method is GET only so pattern is also required
        purl.pattern = '.*';
    }
    var encodedPurl = 
        (purl.methods?purl.methods.join(',')+' ':'') + 
        (purl.pattern?purl.pattern+' ':'') +
        purl.template;
    return encodedPurl;
}

function setupRuleSelects(container, mode, allowedActions) {
    if (mode === 'RewriteRule') {
        container.find('select.requestMethod').selectize();
    }

    if (mode === 'Folder') {
        /*
        container.find('select.actionType').selectize({
            options: actionTypeOptions, // IS THIS NEEDED?
            onFocus: function() { 
                setRemainingOptions(this, actionTypeOptions, $('#rules select.actionType'));
            },
            onInitialize: function() {
                this.addItem('calli:alternate');
            }
        });
        */
    } else {
        if ((mode === 'Redirect' || mode === 'Purl') && allowedActions.length !== 0) {
            if (allowedActions.length === 1) {
                container.find('select.actionType').replaceWith(
                    "<input type='hidden' data-value='"+allowedActions[0].value+"'/>"
                );
                $('#redirectStatus').hide();
            } else {
                var select = container.find('select.actionType').selectize()[0].selectize;
                select.clearOptions();
                select.addOption(allowedActions);
                select.refreshOptions(false);
                if (select.getOption('calli:alternate').length !== 0) {
                    select.addItem('calli:alternate', false);
                } else {
                    select.addItem(allowedActions[0].value, false);
                }
            }
        } else {
            container.find('select.actionType').selectize();
        }
    }
}

function addRule(container, mode) {
    if (!mode) {
        mode = getMode();
    }
    var row = $('#blankRule>div').clone();
    row.find('.purlTarget').after('<input type="hidden" value="" property="calli:alternate"/>');
    $(container).append(row);
    if ($(container).find('.rule').length > 1) {
        $(container).find('.rule').first().find('.glyphicon-minus').show();
    }
    setupRuleSelects($(container).children().last(), mode);
}

function deleteRule(event, container) {
    $(event.target).parents('.rule').remove();
    if ($(container).find('.rule').length === 1) {
        $(container).find('.glyphicon-minus').hide();
    }
}

function parseQueryString() {
    var str = window.location.search;
    if (typeof str !== "string" || str.length === 0) return {};
    var s = str.split("&");
    var s_length = s.length;
    var bit, query = {}, first, second;
    for(var i = 0; i < s_length; i++) {
        bit = s[i].split("=");
        first = decodeURIComponent(bit[0]);
        if (first.length === 0) continue;
        second = decodeURIComponent(bit[1]);
        if(typeof query[first] == "undefined") query[first] = second;
        else if(query[first] instanceof Array) query[first].push(second);
        else query[first] = [query[first], second];
    }
    return query;
}

function removeDefinitions(allowedActions) {
    if (allowedActions.length !== 0) {
        $('#actionDefinitions>dt').each(function() {
            var found = false;
            var action = $(this).text();
            for (var i in allowedActions) {
                if (allowedActions[i].text === action) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                $(this).hide();
                $(this).next().hide();
            }
        });
    }
}